internal/handlers/common/common.go

```diff
@@ -15,10 +15,32 @@
 // Package common provides common code for all handlers.
 package common
 
+import (
+	"github.com/FerretDB/FerretDB/internal/handlers/commonerrors"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/wire"
+)
+
 const (
 	// MinWireVersion is the minimal supported wire protocol version.
 	MinWireVersion = int32(0) // needed for some apps and drivers
 
 	// MaxWireVersion is the maximal supported wire protocol version.
 	MaxWireVersion = int32(17)
 )
+
+// CheckClientMetadata checks if the message does not contain client metadata.
+func CheckClientMetadata(msg *wire.OpMsg) error {
+	document, err := msg.Document()
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	if client, _ := document.Get("client"); client != nil {
+		return commonerrors.NewCommandErrorMsg(commonerrors.ErrClientMetadataCannotBeMutated,
+			"The client metadata document may only be sent in the first hello",
```

Comment:
Okay, so the client metadata may be included in the first `hello` message but not in subsequent. But then we call that function in every `hello` message handler. We don't store the information about it being the first or not call anywhere.

I can't see how that possibly could work, tests notwithstanding.

Reply:
@AlekSi It is a bit unclear to me. Would you be able to draft/explain the scenario?
I would like to observe it on original MongoDB and then to provide next changes.

Thanks in advance!

Reply:
Start FerretDB in the proxy-diff mode (in it the client would receive responses from MongoDB) with `task run-proxy` in one terminal and mongosh with `task mongosh` in another. The very first message `mongosh` sends is:

```json
{
  "Flags": 0,
  "FullCollectionName": "admin.$cmd",
  "NumberToReturn": -1,
  "NumberToSkip": 0,
  "Query": {
    "$k": [
      "ismaster",
      "helloOk",
      "client",
      "compression"
    ],
    "ismaster": 1,
    "helloOk": true,
    "client": {
      "$k": [
        "application",
        "driver",
        "platform",
        "os"
      ],
      "application": {
        "$k": [
          "name"
        ],
        "name": "mongosh 1.10.5"
      },
      "driver": {
        "$k": [
          "name",
          "version"
        ],
        "name": "nodejs|mongosh",
        "version": "5.7.0|1.10.5"
      },
      "platform": "Node.js v16.20.2, LE",
      "os": {
        "$k": [
          "name",
          "architecture",
          "version",
          "type"
        ],
        "name": "linux",
        "architecture": "arm64",
        "version": "5.15.49-linuxkit-pr",
        "type": "Linux"
      }
    },
    "compression": [
      "none"
    ]
  }
}
```
That's an OP_QUERY command that is handled there: https://github.com/FerretDB/FerretDB/blob/main/internal/handlers/pg/cmd_query.go. It can also be sent as OP_MSG `hello` and `isMaster`/`ismaster` command. The `client` field is the client's metadata.

After that initial metadata is sent, all future invocations of one of those commands should not contain the metadata. (The original issue implied it can be sent if it is exactly the same, but I don't think that's true.)
```
db.runCommand({isMaster: 1}) // works

db.runCommand({isMaster: 1, client: {}})
MongoServerError: The client metadata document may only be sent in the first hello
```

So the code should handle all variants of that command, including OP_QUERY, store metadata in `ConnInfo` (or just store that it was received already), and reject future commands with metadata.